node/src/store/utils.rs

```diff
@@ -0,0 +1,548 @@
+            self.0.push(range);
+        } else {
+            self.0[range_index] = range;
+        }
+
+        if let Some(to_remove) = range_to_remove {
+            self.0.remove(to_remove);
+        }
+    }
+
+    /// Return whether range is empty
+    pub fn is_empty(&self) -> bool {
+        self.0.iter().all(|r| r.is_empty())
+    }
+
+    pub fn contains(&self, height: u64) -> bool {
+        self.0.iter().any(|r| r.contains(&height))
+    }
+
+    /// Return highest height in the range
+    pub fn head(&self) -> Option<u64> {
+        self.0.last().map(|r| *r.end())
+    }
+
+    /// Return lowest height in the range
+    pub fn tail(&self) -> Option<u64> {
+        self.0.first().map(|r| *r.start())
+    }
+}
+
+impl Display for HeaderRanges {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "[")?;
+        for (idx, range) in self.0.iter().enumerate() {
+            if idx == 0 {
+                write!(f, "{}..={}", range.start(), range.end())?;
+            } else {
+                write!(f, ", {}..={}", range.start(), range.end())?;
+            }
+        }
+        write!(f, "]")
+    }
+}
+
+impl AsRef<[RangeInclusive<u64>]> for HeaderRanges {
+    fn as_ref(&self) -> &[RangeInclusive<u64>] {
+        &self.0
+    }
+}
+
+impl FromIterator<RangeInclusive<u64>> for HeaderRanges {
+    fn from_iter<T: IntoIterator<Item = RangeInclusive<u64>>>(iter: T) -> Self {
+        Self(iter.into_iter().collect())
+    }
+}
+
+impl<const T: usize> From<[RangeInclusive<u64>; T]> for HeaderRanges {
+    fn from(value: [RangeInclusive<u64>; T]) -> Self {
+        Self(value.into_iter().collect())
+    }
+}
+
+impl IntoIterator for HeaderRanges {
+    type Item = u64;
+    type IntoIter = HeaderRangesIterator;
+    fn into_iter(self) -> Self::IntoIter {
+        let mut outer_iter = self.0.into_iter();
+        HeaderRangesIterator {
+            inner_iter: outer_iter.next(),
+            outer_iter,
+        }
+    }
+}
+
+pub struct HeaderRangesIterator {
+    inner_iter: Option<RangeInclusive<u64>>,
+    outer_iter: IntoIter<[RangeInclusive<u64>; 2]>,
+}
+
+impl HeaderRangesIterator {
+    pub fn next_batch(&mut self, limit: u64) -> Option<RangeInclusive<u64>> {
+        let current_range = self.inner_iter.take()?;
+
+        if current_range.len() <= limit {
+            self.inner_iter = self.outer_iter.next();
+            Some(current_range)
+        } else {
+            let returned_range = *current_range.start()..=*current_range.start() + limit - 1;
+            self.inner_iter = Some(*current_range.start() + limit..=*current_range.end());
+            Some(returned_range)
+        }
+    }
+}
+
+impl Iterator for HeaderRangesIterator {
+    type Item = u64;
+    fn next(&mut self) -> Option<Self::Item> {
+        if let Some(v) = self.inner_iter.as_mut()?.next() {
+            return Some(v);
+        }
+        self.inner_iter = self.outer_iter.next();
+        self.next()
```

Comment:
I prefer loop instead of recursion in this case.

```suggestion
        loop {
            if let Some(v) = self.inner_iter.as_mut()?.next() {
                return Some(v);
            }
            self.inner_iter = self.outer_iter.next();
        }
```